tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,214 @@
+
+#include "tc/autotuner/genetic_search.h"
+#include "tc/autotuner/options_cache.h"
+#include "tc/autotuner/parameters.h"
+#include "tc/autotuner/utils.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/parser.h"
+
+namespace tc {
+namespace autotune {
+
+namespace detail {
+/**
+ * Internal harness to support multithreaded compilation and evaluation over
+ * various backends for a particular SearchStrategy.
+ * The SearchStrategy is a template parameter passed to the run method.
+ */
+template <typename Backend>
+class TuningHarness {
+ public:
+  using BackendType = Backend;
+  using ExecutorType = typename BackendType::ExecutorType;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+  using OptionsCacheType = OptionsCache<Backend>;
+
+  TuningHarness(
+      size_t maxPopulationSize,
+      lang::TreeRef kTcTree,
+      const std::unordered_map<size_t, std::vector<const DLConstTensor*>>&
+          inputs,
+      std::unordered_map<size_t, std::vector<const DLTensor*>>& outputs,
+      const typename Backend::MappingOptionsType& baseMapping,
+      const TuningParameterFixer& fixedParams,
+      std::shared_ptr<OptionsCache<Backend>>& optionsCache);
+
+  /// Runs a SearchStrategy over numIterations
+  template <typename SearchStrategy>
+  void run(SearchStrategy* searchStrategy);
+
+  /// Once a signal has been caught, a flag is set and we terminate after the
+  /// current iteration is finished. This is apparently necessary for proper
+  /// python termination.
+  /// TODO: we should detect when we come from python and exit properly in C++.
+  void stopAfterCurrentIteration();
+
+ private:
+  /// Traverse one iteration of candidates in parallel and evaluate their
+  /// runtimes
+  template <typename SearchStrategy>
+  void runOneIteration(SearchStrategy* searchStrategy, size_t iteration);
+
+  /// Helper function to delegate compiling on the cpu to different threads
+  template <typename SearchStrategy>
+  void doCompile(SearchStrategy* searchStrategy);
+
+  /// Helper function to get a kernel into benchmark-able state
+  /// This function must be specialized per Backend
+  /// TODO: virtual inheritance?
+  bool warmupOrPrune(
+      ExecutorType& executor,
+      const std::vector<const DLTensor*>& outputs,
+      const std::vector<const DLConstTensor*>& inputs,
+      size_t bestTimeSoFar);
+
+  /// Helper function to delegate running the compiled code to different
+  /// threads.
+  /// This function must be specialized per Backend
+  /// TODO: virtual inheritance?
+  void doEvaluate(size_t device, size_t populationSize, Printer& printer);
+
+  const MappingOptionsType& bestMappingOption() {
+    std::lock_guard<std::mutex> lock(bestTimeMtx_);
+    return bestMappingOptions_;
+  }
+
+ private:
+  static constexpr int kReducedWarmupIterations = 2;
+  static constexpr int kReducedBenchmarkIterations = 10;
+  static constexpr int kEarlyPruneFactor = 5;
```

Comment:
Please add some explanation.
The meaning of these constants may be obvious from their names to you,
but not to me.